### ISSUE TITLE
failure to infer move mode for a variable when the move is from `*x`

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#23675]
Example (play: http://is.gd/idpRaW)

``` rust
use std::mem;
struct Trace;
fn foo(b: Box<Trace>) {
    do1(|| mem::drop(*b));
}
fn do1<F:FnOnce()>(f: F) { }
fn main() { }
```

If you change `mem::drop(*b)` to `mem::drop(b)`, it compiles.

**Comment by bstrie** [comment]
Triage: still relevant.

**Comment by memoryruins** [comment]
Triage: now compiles on both editions.
`rustc: 1.32.0`

**Comment by matthewjasper** [comment]
The fix appears to have been in `1.23`
